src/graphql_relay/connection/array_connection.py

```diff
@@ -232,3 +242,180 @@ def get_offset_with_default(
+
+    return (
+        edges,
+        has_previous_page,
+        has_next_page,
+    )
+
+
+def _handle_last_before(
+    array_slice: SizedSliceable,
+    array_slice_length: Optional[int],
+    array_length: Optional[int],
+    last: Optional[int],
+    before: Optional[str],
+    slice_start: int = 0,
+    edge_type: type[Edge] = Edge,
+) -> tuple[list[Edge], bool, bool]:
+    """Handle the `last` and `before` arguments."""
+
+    if last is not None and last < 0:
+        raise ValueError("Argument 'last' must be a non-negative integer.")
+
+    # If defined, convert `before` cursor into an offset.
+    before_offset: Optional[int] = cursor_to_offset(before) if before else None
+
+    # If the length of the array is not provided, calculate it.
+    if array_length is None:
+        array_length = _calculate_array_length(array_slice, array_slice_length)
+
+    # Calculate the `end_offset`:
+    # If `before` is provided, use it as `end_offset` (cropping it to the bounds of the slice).
+    # Otherwise, the `end_offset` is the end of the slice.
+    end_offset: int
+    if before_offset is not None and before_offset >= 0:
+        end_offset = min(before_offset, array_length)
+
+    else:
+        end_offset = array_length
+
+    # Calculate the `start_offset`:
+    # If `last` is provided, use it to calculate the `start_offset` by subtracting it from the `end_offset`,
+    # ensuring that it is greater than or equal to zero, or to the start of slice (whichever is greater).
+    # Otherwise, the `start_offset` is the start of the slice.
+    start_offset: int
+    if last is not None:
+        start_offset = max(end_offset - last, max(slice_start, 0))
+
+    else:
+        start_offset = 0
+
+    trimmed_slice: SizedSliceable = array_slice[
+        start_offset - slice_start : end_offset - slice_start
+    ]
+
+    has_previous_page: bool = start_offset > 0
+
+    has_next_page: bool = end_offset < array_length
+    if before_offset is not None and before_offset < 0:
+        has_next_page = True
+
+    edges: list[Edge] = [
+        edge_type(
+            node=node,
+            cursor=offset_to_cursor(start_offset + index),
+        )
+        for index, node in enumerate(trimmed_slice)
+    ]
+
+    return (
+        edges,
+        has_previous_page,
+        has_next_page,
+    )
+
+
+def _calculate_array_length(
+    array_slice: SizedSliceable,
+    array_slice_length: Optional[int],
+) -> int:
+    """Calculate the length of the array.
+
+    If `array_slice_length` is provided, use it.
+    Otherwise, if `array_slice` is a list, use `len(array_slice)`.
+    Otherwise, if `array_slice` has a `count` method, use `array_slice.count()`.
+    Otherwise, use `len(array_slice)`."""
+    if array_slice_length is not None:
+        return array_slice_length
+
+    if isinstance(array_slice, list):
+        return len(array_slice)
+
+    if hasattr(array_slice, "count"):
+        return array_slice.count()
+
+    return len(array_slice)
```

Comment:
Same result but a bit less verbose? 

```suggestion
def _calculate_array_length(
    array_slice: SizedSliceable,
    array_slice_length: Optional[int],
) -> int:
    """Calculate the length of the array.

    If `array_slice_length` is provided, use it.
    Otherwise, if `array_slice` has a `count` method, use `array_slice.count()`.
    Otherwise, use `len(array_slice)`."""
    if array_slice_length is not None:
        return array_slice_length

    if hasattr(array_slice, "count"):
        return array_slice.count()

    return len(array_slice)
```

Reply:
Actually no because `list` has a `count` method that means something else.